### PR TITLE
ENH: Add Sidi's zero finder as optimize.zeros.sidi

### PR DESCRIPTION
**Pull request by pvanmulbregt**
Added a function to optimize.zeros.py
`sidi(f, a, b, args=(), k=2, xtol=_xtol, rtol=_rtol, maxiter=_iter, is_bracket=True, full_output=False, disp=True)`

Sidi's root-finder is a family of generalizations of the secant method using Newton's interpolation formula with divided differences to construct a new approximation to a root of a function.

Muller's method use the interpolating polynomial `p_k` to approximate `f` from `k+1` points.  Sidi's method uses the derivative of the interpolating polynomial `p_k'`  to approximate the derivative `f'` of `f`. The generated approximation is then (reminiscent of Newton-Raphson)
```
        f(x0)
x0 -  --------
       p_k'(x0)
```

For `k=1`, this reduces to the Secant method, with approximate order of convergence `1.618`.
For `k=2, 3,...`, the approximate order of convergence is `1.839, 1.928,...`, a sequence approaching 2, whenever applied to functions with `k+1` continuous derivatives on the interval.

Using divided differences, the method quickly computes ` p_k'(x0)` and only involves evaluation of the function `f` itself, not any derivatives of `f`.
One potential advantage over the Newton-Raphson routine `optimize.newton`, is that it can require fewer evaluations of `f` and `f'`.  One potential advantage over methods such as `optimize.brentq` is that by incorporating 2nd and higher order information it can lead to fewer iterations and function evaluations.

Most of the function's arguments are the same as for `bisect/brentq/ridder`.  There are two additional arguments:
`k`: the order of the interpolating polynomial.  Default is 2.
`is_bracket`: If False, `a` and `b` are just the first two iterates  and not necessarily an interval containing a root.  Similar to Newton-Raphson or Secant, Sidi's method does not require a bracketing interval for operation. If a bracketing interval is provided then it will used to constrain the iterations.

Tests have been updated to include testing of `optimize.sidi`.

I also added keyword args `full_output=False, disp=True` to `optimize.newton(...)` to better match the other zero finders.  The default values preserve the current behavior. 
Setting `full_output=True` causes a `RootResults` object to be returned with information on the total number of function_calls and iterations. 
Setting `disp=False` suppresses the RuntimeError associated with non-convergence, and is likely best combined with `full_output=True`.

### CODE REVIEW

File: scipy/optimize/tests/test_zeros.py [hunk 1]
```diff
@@ -39,6 +42,39 @@ def test_brentq(self):
     def test_brenth(self):
         self.run_check(cc.brenth, 'brenth')
 
+    def test_sidi_k1(self):
```
Comment (tylerjereddy) [inline scipy/optimize/tests/test_zeros.py]: Could perhaps gain some concision by using a parametrized pytest for the value of `k` and just have one single `test_sidi_k` test.

**Comment by pvanmulbregt** [comment]
My thinking on Sidi's root-finder in SciPy has evolved.  It may be better to introduce it as a non-bracketed algorithm which would also allow complex-valued functions.  So I'm going to split this into a couple of PRs: gh-8803 for the Newton improvements and slight additional infrastructure, and a separate PR for Sidi's algorithm.

File: scipy/optimize/tests/test_zeros.py (hunk 1 above)
Comment (pvanmulbregt) [reply scipy/optimize/tests/test_zeros.py]: How is that specified for member functions? https://docs.pytest.org/en/2.9.0/parametrize.html only shows it for stand-alone functions.  [Separate question is then whether that functionality is supported in all the versions of pytest currently supported by SciPy...]

**Comment by rgommers** [comment]
@pvanmulbregt gh-8803 is merged. Did you intend to keep this PR open or replace it?

**Comment by pvanmulbregt** [comment]
I'm closing it.  gh-8803 and a separate PR for Sidi's method when I'm happy with it.

**Pull request closed without merging**